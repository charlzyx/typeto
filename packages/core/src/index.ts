export { debugInfo, debugNode, debugSymbol, debugType } from "./debug";
export { getNodeExtraInfo, getPropNodeTypeRefName } from "./getNodeExtraInfo";
export { TypeResolver } from "./TypeResolver";
