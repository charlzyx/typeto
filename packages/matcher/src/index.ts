export * from "./AstMatcher";
export * from "./matchers";
