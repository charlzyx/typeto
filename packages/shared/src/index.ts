// export { debugInfo, debugNode, debugSymbol, debugType } from "./debug";
export { getNodeExtraInfo } from "./getNodeExtraInfo";
