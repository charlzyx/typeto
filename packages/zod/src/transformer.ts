import { ClassDeclaration, Node, Type, YieldExpression } from "ts-morph";
import {
  TypeResolver,
  getNodeExtraInfo,
  getPropNodeTypeRefName,
} from "@typeto/core";

interface ResolveContext {
  globalRefs: WeakMap<Type, string>;
  selfCircularRefs: WeakMap<Type, string>;
  touched: { current: boolean };
  self: string;
}

const zodResolver = new TypeResolver<string, ResolveContext>()
  .before((type, ctx, resolver) => {
    // const itis = debugType(type as any);
    // 检测自身引用
    if (ctx.selfCircularRefs.has(type)) {
      const has = ctx.selfCircularRefs.get(type);
      ctx.touched.current = true;
      return has;
    }

    // 检测循环引用
    if (ctx.globalRefs.has(type)) {
      const has = ctx.globalRefs.get(type);
      if (has !== ctx.self) return has;
    }
  })
  .null(() => `z.null()`)
  .undefined(() => `z.undefined()`)
  .stringLiteral(
    (type) => `z.literal(${JSON.stringify(type.getLiteralValue())})`
  )
  .numberLiteral((type) => `z.literal(${type.getLiteralValue()})`)
  .booleanLiteral((type) => `z.literal(${type.getText()})`)
  .string((type) => {
    return `z.string()`;
  })
  .boolean(() => `z.boolean()`)
  .number(() => `z.number()`)
  .literal((type) => `z.literal(${JSON.stringify(type.getLiteralValue())})`)
  .intersection((type, ctx, resovler) => {
    const intersectionTypes = type
      .getIntersectionTypes()
      .map((t) => resovler.resolve(t, ctx))
      .join(", ");
    return `z.intersection([${intersectionTypes}])`;
  })
  .union((type, ctx, resolver) => {
    const unionTypes = type.getUnionTypes();
    const subTypes = unionTypes.map((t) => resolver.resolve(t, ctx));

    // 尝试检测是否为判别联合
    const discriminator = subTypes.find((x) => /object/.test(x));
    const prefix = discriminator ? "z.discriminatedUnion" : "z.union";
    // 普通联合类型
    return `${prefix}([${subTypes.join(", ")}])`;
  })
  .array((type, ctx, resovler) => {
    return `z.array(${resovler.resolve(type.getArrayElementType(), ctx)})`;
  })
  .object((type, ctx, resolver) => {
    ctx.globalRefs.set(type, `${type.getText()}`);
    ctx.selfCircularRefs.set(type, `${type.getText()}`);

    const properties = type
      .getProperties()
      .map((propSymbol) => {
        // 获取属性的声明节点

        const propNode =
          propSymbol.getValueDeclaration() ?? propSymbol.getDeclarations()[0];

        const extra = getNodeExtraInfo(propNode);

        const message = extra.info["message"] || extra.comment;

        // 获取属性的类型
        const propType = propNode
          ? propSymbol.getTypeAtLocation(propNode)
          : propSymbol.getDeclaredType();
        if (!propType) return "";

        let zodType = resolver.resolve(propType, ctx);
        const stringSpecific = getPropNodeTypeRefName(propNode, true);

        // 处理可选属性
        const isOptional = propSymbol.isOptional();
        // 检查是否为 nullish (null 或 undefined)
        const isNullable = propType.isNull() || propType.isNullable();
        if (message) {
          zodType = zodType.replace("()", `({ message: "${message}"})`);
        }
        if (stringSpecific) {
          zodType = `${zodType}.${stringSpecific}()`;
        }
        if (isNullable) {
          zodType = `${zodType}.nullable()`;
        }
        if (isOptional) {
          zodType = `${zodType}.optional()`;
        }

        return `${propSymbol.getName()}: ${zodType}`;
      })
      .filter(Boolean)
      .join(",\n  ");

    return `z.object({\n  ${properties}\n})`;
  });

export const transformDefinitions = (
  definitions: ClassDeclaration[]
): string => {
  const globalRefs = new WeakMap<Type, string>();

  const imports = `import { z } from "zod";\n\n`;

  const schemas = definitions
    .map((def) => {
      const className = def.getName();
      const selfCircularRefs = new WeakMap<Type, string>();
      const touched = { current: false };

      const schema = zodResolver.resolve(def.getType(), {
        globalRefs,
        selfCircularRefs,
        touched,
        self: def.getName(),
      });

      if (touched.current) {
        return `export const ${className}: z.ZodSchema<${className}> = z.lazy(() =>  ${schema})`;
      } else {
        return `export const ${className} = ${schema};`;
      }
    })
    .join("\n\n");

  return imports + schemas;
};
