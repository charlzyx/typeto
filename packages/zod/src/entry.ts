import { ClassDeclaration, Node, Project } from "ts-morph";
import { getNodeExtraInfo } from "@typeto/shared";

// 忽略规则的键
const IgnoreRuleKeys = {
  jsDoc: ["ignore", "WIP", "Draft"],
  classDecorators: ["ignore", "WIP", "Draft"],
  leadingComment: ["@ignore", "@WIP", "@Draft"],
};

// 判断是否应跳过该节点
const shouldSkip = (node: ClassDeclaration) => {
  const { decorators, jsDocs, leadingComments } = getNodeExtraInfo(node);

  // 检查装饰器中是否有忽略标签
  const skipByDecorator =
    Object.keys(decorators).findIndex((decoratorName) =>
      IgnoreRuleKeys.classDecorators.includes(decoratorName)
    ) > -1;
  if (skipByDecorator) return true;

  // 检查 jsDoc 中是否有忽略标签
  const skipByJsDoc =
    Object.keys(jsDocs).findIndex((tagName) =>
      IgnoreRuleKeys.jsDoc.includes(tagName)
    ) > -1;
  if (skipByJsDoc) return true;

  // 检查前导注释中是否有忽略标签
  const skipByComment =
    leadingComments
      .join("\n")
      .split(/\s+/)
      .findIndex((word) => IgnoreRuleKeys.leadingComment.includes(word)) > -1;
  if (skipByComment) return true;

  return false;
};

export const getEntryNodes = (project: Project) => {
  const definitions: ClassDeclaration[] = [];

  project.getSourceFiles().forEach((sourceFile) => {
    sourceFile.getClasses().forEach((classDec) => {
      if (!shouldSkip(classDec)) {
        definitions.push(classDec);
      }
    });
  });

  return { definitions };
};
