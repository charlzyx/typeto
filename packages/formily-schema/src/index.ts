import { Project } from "ts-morph";
import { transformDefinitions } from "./transformer";
import { getEntryNodes } from "./entry";
import { ISchema } from "@formily/json-schema";

const clean = (schema: ISchema) => {
  const replacer = (k: string, v: any) => {
    if (k === "x-code") return undefined;
    return v;
  };

  return JSON.parse(JSON.stringify(schema, replacer));
};

export const transform = (project: Project) => {
  const { definitions } = getEntryNodes(project);
  const schema = transformDefinitions(definitions);
  const schemaDefs = clean(schema);
  const prefix = 'import { ISchema } from "@formily/json-schema";';
  const reforamt = Object.keys(schemaDefs).map((def) => {
    return `const ${def}: ISchema = ${JSON.stringify(
      schemaDefs[def],
      null,
      2
    )};\n`.replace(/"#\w+"/g, (m) => {
      return m.replace(/"/g, "").replace("#", "");
    });
  });
  return [prefix, ...reforamt].join("\n");
};
