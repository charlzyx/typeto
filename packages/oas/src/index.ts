import { Project } from "ts-morph";
// import { transformDefinitions, transformOperations } from "./transformer";
import { transformDefinitions, transformOperations } from "./transformer2";
import { getEntryNodes } from "./entry";

const clean = (x: any) =>
  JSON.parse(
    JSON.stringify(x, (k: string, v) => {
      if (k === "_code") return undefined;
      return v;
    })
  );

export const transform = (project: Project) => {
  const { definitions, operations } = getEntryNodes(project);
  const { defSchema, defNameMap } = transformDefinitions(definitions);
  const opSchema = transformOperations(operations, defNameMap);
  return { operations: clean(opSchema), definitions: clean(defSchema) };
};
