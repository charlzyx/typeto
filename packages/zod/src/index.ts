import { Project } from "ts-morph";
import { transformDefinitions } from "./transformer";
import { getEntryNodes } from "./entry";

export const transform = (project: Project) => {
  const { definitions } = getEntryNodes(project);
  const schema = transformDefinitions(definitions);
  return schema;
};
