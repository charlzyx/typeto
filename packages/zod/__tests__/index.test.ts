import { describe, expect, it } from "bun:test";
import { Project } from "ts-morph";
import { transform } from "../src";

describe("zod schema tests", () => {
  const project = new Project({});

  it("复杂组合测试 / should handle complex ", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
type int32 = number;
type int64 = number;
type float = number;
type double = number;
type date = string;
type datetime = string;
type password = string;
type binary = string;
type byte = string;

type Status = "placed" | "approved" | "delivered";

class Order {
  id: int64;
  petId: int64;
  quantity: int32;
  shipDate: datetime;
  status: Status;
  complete: boolean;
}

class Customer {
  id: int64;
  username: string;
  address: Address;
}

class Address {
  street: string;
  city: string;
  state: string;
  zip: number;
}

class Category {
  id: int64;
  name: string;
}

class User {
  id: int64;
  username: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  phone: string;
  userStatus: int32;
}

class Tag {
  id: int64;
  name: string;
}

class Pet {
  /** Pet's id */
  id: int64 = 0;
  // getJsDocs() cannot see me
  category: Category; // but getLeadingCommentRanges/getTrailingCommentRanges can do it!
  // name with default
  name: string = "hi";
  /** photos */
  photoUrls: string[];
  tags: Tag[];
  status: Status;
}

class TreeLike {
    children: TreeLike[]
}

      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理基本类型 / should handle basic types correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      class BasicTypes {
        str: string;
        num: number;
        bool: boolean;
        nullValue: null;
        undefinedValue: undefined;
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理字面量类型 / should handle literal types correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      class LiteralTypes {
        strLiteral: "hello";
        numLiteral: 42;
        boolLiteral: true;
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理联合类型 / should handle union types correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      class User {
        status: "active" | "inactive";
        role: "admin" | "user" | "guest";
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理判别联合 / should handle discriminated unions correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      type Shape =
        | { kind: "circle"; radius: number }
        | { kind: "square"; sideLength: number };

      class Drawing {
        shape: Shape;
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();

    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理交叉类型 / should handle intersection types correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      interface Named { name: string }
      interface Aged { age: number }
      
      class Person {
        info: Named & Aged;
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();

    project.removeSourceFile(sourceFile);
  });

  // it("应该正确处理可选和可空属性 / should handle optional and nullable properties correctly", () => {
  //   /**
  //    * TODO: 有个小问题
  //    * https://ts-ast-viewer.com/#code/MYGwhgzhAECqEFMBO0DeAoaXoDswFsEAuaCAFyQEscBzAbk2zBoQH4ScBXfAI2QezQelAPYlyVWtAA+uTiBACmAE2VIEUdqQrUaMuQoYBfIA
  //    * string | null 节点在 compiler api 中使用 getTypeAtLocation 获取到的节点是 string 类型
  //    * 而不是 UnionType , 但是 online 页面是对的
  //    *
  //    */
  //   const sourceFile = project.createSourceFile(
  //     "test.ts",
  //     `
  //     class User {
  //       name: string;
  //       age?: number;
  //       bio: string | null;
  //       address?: string | null;
  //     }
  //     `
  //   );

  //   const result = transform(project);
  //   expect(result).toContain("name: z.string()");
  //   expect(result).toContain("age: z.number().optional()");
  //   expect(result).toContain("bio: z.string().nullable()");
  //   expect(result).toContain("address: z.string().nullable().optional()");

  //   project.removeSourceFile(sourceFile);
  // });

  it("应该正确处理嵌套对象和数组 / should handle nested objects and arrays correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      class Address {
        street: string;
        city: string;
      }
      
      class User {
        name: string;
        address: Address;
        tags: string[];
        contacts: Address[];
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();

    project.removeSourceFile(sourceFile);
  });

  it("应该正确处理循环引用 / should handle circular references correctly", () => {
    const sourceFile = project.createSourceFile(
      "test.ts",
      `
      class Employee {
        name: string;
        // 直接循环引用
        supervisor?: Employee;
        // 嵌套循环引用
        team?: {
          leader: Employee;
          members: Employee[];
        };
      }
      `
    );

    const result = transform(project);
    expect(result).toMatchSnapshot();
    project.removeSourceFile(sourceFile);
  });
});
