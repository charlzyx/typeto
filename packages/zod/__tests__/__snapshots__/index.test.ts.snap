// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`zod schema tests 复杂组合测试 / should handle complex  1`] = `
"import { z } from "zod";

export const Order = z.object({
  id: z.number().int64(),
  petId: z.number().int64(),
  quantity: z.number().int32(),
  shipDate: z.string().datetime(),
  status: z.union([z.literal("placed"), z.literal("approved"), z.literal("delivered")]),
  complete: z.boolean()
});

export const Customer = z.object({
  id: z.number().int64(),
  username: z.string(),
  address: z.object({
  street: z.string(),
  city: z.string(),
  state: z.string(),
  zip: z.number()
})
});

export const Address = z.object({
  street: z.string(),
  city: z.string(),
  state: z.string(),
  zip: z.number()
});

export const Category = z.object({
  id: z.number().int64(),
  name: z.string()
});

export const User = z.object({
  id: z.number().int64(),
  username: z.string(),
  firstName: z.string(),
  lastName: z.string(),
  email: z.string(),
  password: z.string(),
  phone: z.string(),
  userStatus: z.number().int32()
});

export const Tag = z.object({
  id: z.number().int64(),
  name: z.string()
});

export const Pet = z.object({
  id: z.number({ message: "Pet's id"}).int64(),
  category: Category,
  name: z.string({ message: "name with default"}),
  photoUrls: z.array(z.string({ message: "photos"})),
  tags: z.array(Tag),
  status: z.union([z.literal("placed"), z.literal("approved"), z.literal("delivered")])
});

export const TreeLike: z.ZodSchema<TreeLike> = z.lazy(() =>  z.object({
  children: z.array(TreeLike)
}))"
`;

exports[`zod schema tests 应该正确处理基本类型 / should handle basic types correctly 1`] = `
"import { z } from "zod";

export const BasicTypes = z.object({
  str: z.string(),
  num: z.number(),
  bool: z.boolean(),
  nullValue: z.null().nullable(),
  undefinedValue: z.undefined()
});"
`;

exports[`zod schema tests 应该正确处理字面量类型 / should handle literal types correctly 1`] = `
"import { z } from "zod";

export const LiteralTypes = z.object({
  strLiteral: z.literal("hello"),
  numLiteral: z.literal(42),
  boolLiteral: z.literal(true)
});"
`;

exports[`zod schema tests 应该正确处理联合类型 / should handle union types correctly 1`] = `
"import { z } from "zod";

export const User = z.object({
  status: z.union([z.literal("active"), z.literal("inactive")]),
  role: z.union([z.literal("admin"), z.literal("user"), z.literal("guest")])
});"
`;

exports[`zod schema tests 应该正确处理判别联合 / should handle discriminated unions correctly 1`] = `
"import { z } from "zod";

export const Drawing = z.object({
  shape: z.discriminatedUnion([z.object({
  kind: z.literal("circle"),
  radius: z.number()
}), z.object({
  kind: z.literal("square"),
  sideLength: z.number()
})])
});"
`;

exports[`zod schema tests 应该正确处理交叉类型 / should handle intersection types correctly 1`] = `
"import { z } from "zod";

export const Person = z.object({
  info: z.intersection([z.object({
  name: z.string()
}), z.object({
  age: z.number()
})])
});"
`;

exports[`zod schema tests 应该正确处理嵌套对象和数组 / should handle nested objects and arrays correctly 1`] = `
"import { z } from "zod";

export const Address = z.object({
  street: z.string(),
  city: z.string()
});

export const User = z.object({
  name: z.string(),
  address: Address,
  tags: z.array(z.string()),
  contacts: z.array(Address)
});"
`;

exports[`zod schema tests 应该正确处理循环引用 / should handle circular references correctly 1`] = `
"import { z } from "zod";

export const Employee: z.ZodSchema<Employee> = z.lazy(() =>  z.object({
  name: z.string(),
  supervisor: Employee.optional(),
  team: z.object({
  leader: Employee,
  members: z.array(Employee)
}).optional()
}))"
`;
